runtime-sdk/src/modules/core/mod.rs

```diff
@@ -453,22 +468,89 @@ impl<Cfg: Config> Module<Cfg> {
-                // how much gas it uses while it fails.
-                let gas_used = *tx_ctx.value::<u64>(CONTEXT_KEY_GAS_USED).or_default();
-                Ok(gas_used + extra_gas)
+        // Simulates a transaction.
+        let mut simulate = |tx: transaction::Transaction, report_failure: bool| {
+            ctx.with_simulation(|mut sim_ctx| {
+                sim_ctx.with_tx(0 /* index */, tx_size, tx, |mut tx_ctx, call| {
+                    let (result, _) =
+                        dispatcher::Dispatcher::<C::Runtime>::dispatch_tx_call(&mut tx_ctx, call);
+                    if !result.is_success() && report_failure {
+                        // Report failure.
+                        let err: TxSimulationFailure = result.try_into().unwrap(); // Guaranteed to be a Failed CallResult.
+                        return Err(Error::TxSimulationFailed(err));
+                    }
+                    // Don't report success or failure. If the call fails, we still report
+                    // how much gas it uses while it fails.
+                    let gas_used = *tx_ctx.value::<u64>(CONTEXT_KEY_GAS_USED).or_default();
+                    Ok(gas_used)
+                })
             })
-        })
+        };
+
+        // Do a binary search for exact gas limit.
+        // If binary search is disabled the loop exits before doing the first iteration.
+        let (mut lo, mut hi) = (10_u128, args.tx.auth_info.fee.gas as u128); // Use u128 to avoid overflows when computing the mid point.
+        let (cap, mut iters, mut first_success) = (hi, 0, false);
+        while (lo + 1 < hi) && iters < bs_max_iters {
+            iters += 1;
+
+            let mid = (hi + lo) / 2;
+            args.tx.auth_info.fee.gas = mid as u64; // hi,lo are capped at Max::u64.
+            match simulate(args.tx.clone(), true) {
+                Ok(r) => {
+                    // Estimate success. Try with lower gas.
+                    hi = mid;
+
+                    // In vast majority of cases the initially returned gas estimate might already be a good one.
+                    // Check if this is the case to greatly speed up the convergence.
+                    if !first_success && (lo + 1 < hi) {
+                        first_success = true;
+
+                        // Check if simulate with the returned estimate succeeds, so we can further clamp the high value.
+                        args.tx.auth_info.fee.gas = r;
+                        match simulate(args.tx.clone(), true) {
+                            Ok(_) => hi = r as u128,
+                            _ => continue,
+                        }
+                        // Check if the returned estimate is also the lower bound. In that case we can exit early.
+                        args.tx.auth_info.fee.gas = (hi - 1) as u64;
+                        match simulate(args.tx.clone(), true) {
+                            Err(_) => lo = hi - 1,
```

Comment:
The early exit is due to `lo + 1 < hi` not holding as `lo + 1 = hi` after this. To make it more clear could either add a comment here or add an explicit loop break.